### PR TITLE
first version of StreamPractice

### PR DESCRIPTION
**Pull request by oleksiistepaniak**

### CODE REVIEW

File: src/main/java/practice/CandidateValidator.java [hunk 1]
```diff
@@ -1,5 +1,26 @@
...
+
+public class CandidateValidator implements Predicate<Candidate> {
+    private static final int MIN_AGE = 35;
+    private static final String NATIONALITY_FOR_ELECTION = "Ukrainian";
+    private static final int INDEX_FOR_FIRST_DATE = 0;
+    private static final int INDEX_FOR_SECOND_DATE = 1;
+    private static final int MIN_NUMBER_OF_LIVING_IN_UKR = 10;
+
+    public CandidateValidator() {
+    }
```
Comment (copypasterr) [inline src/main/java/practice/CandidateValidator.java:14]: This constructor is needed here?

File: src/main/java/practice/CandidateValidator.java [hunk 2]
```diff
@@ -1,5 +1,26 @@
...
+
+    @Override
+    public boolean test(Candidate candidate) {
+        int firstDate = Integer.parseInt(candidate.getPeriodsInUkr()
+                .split("-")[INDEX_FOR_FIRST_DATE]);
+        int secondDate = Integer.parseInt(candidate.getPeriodsInUkr()
+                .split("-")[INDEX_FOR_SECOND_DATE]);
+        return candidate.isAllowedToVote() && candidate.getAge() >= MIN_AGE
+                && candidate.getNationality().equals(NATIONALITY_FOR_ELECTION)
+                && (secondDate - firstDate) >= MIN_NUMBER_OF_LIVING_IN_UKR;
```
Comment (copypasterr) [inline src/main/java/practice/CandidateValidator.java:24]: You can use boolean checker methods for more readable code

**Review by copypasterr — APPROVED** [review]

File: src/main/java/practice/CandidateValidator.java [hunk 3]
```diff
@@ -1,5 +1,26 @@
...
+    private static final int INDEX_FOR_SECOND_DATE = 1;
+    private static final int MIN_NUMBER_OF_LIVING_IN_UKR = 10;
+
+    public CandidateValidator() {
+    }
+
+    @Override
+    public boolean test(Candidate candidate) {
+        int firstDate = Integer.parseInt(candidate.getPeriodsInUkr()
+                .split("-")[INDEX_FOR_FIRST_DATE]);
```
Comment (sarakhmen) [inline src/main/java/practice/CandidateValidator.java:19]: I'd recommend introducing a local variable for candidate.getPeriodsInUkr().split("-")

File: src/main/java/practice/CandidateValidator.java (hunk 2 above)
Comment (sarakhmen) [inline src/main/java/practice/CandidateValidator.java:24]:
```suggestion
                && secondDate - firstDate >= MIN_NUMBER_OF_LIVING_IN_UKR;
```

File: src/main/java/practice/StreamPractice.java [hunk 4]
```diff
@@ -50,7 +67,11 @@ public List<Person> selectMenByAge(List<Person> peopleList, int fromAge, int toA
      */
     public List<Person> getWorkablePeople(int fromAge, int femaleToAge,
                                           int maleToAge, List<Person> peopleList) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(p -> (p.getSex().equals(Person.Sex.WOMAN) && p.getAge() >= fromAge
+                        && p.getAge() <= femaleToAge) || (p.getSex().equals(Person.Sex.MAN)
+                        && p.getAge() >= fromAge && p.getAge() <= maleToAge))
```
Comment (sarakhmen) [inline src/main/java/practice/StreamPractice.java:73]: As stated in Common Mistakes we can compare Enums using '==' operator

File: src/main/java/practice/StreamPractice.java [hunk 5]
```diff
@@ -35,7 +49,10 @@ public Double getOddNumsAverage(List<Integer> numbers) {
      * Example: select men who can be recruited to army (from 18 to 27 years old inclusively).
      */
     public List<Person> selectMenByAge(List<Person> peopleList, int fromAge, int toAge) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(p -> p.getSex().equals(Person.Sex.MAN)
```
Comment (sarakhmen) [inline src/main/java/practice/StreamPractice.java:53]: The same issue with Enums comparison

File: src/main/java/practice/StreamPractice.java [hunk 6]
```diff
@@ -59,7 +80,12 @@ public List<Person> getWorkablePeople(int fromAge, int femaleToAge,
      * return the names of all cats whose owners are women from `femaleAge` years old inclusively.
      */
     public List<String> getCatsNames(List<Person> peopleList, int femaleAge) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(p -> p.getSex().equals(Person.Sex.WOMAN) && p.getAge() >= femaleAge)
```
Comment (sarakhmen) [inline src/main/java/practice/StreamPractice.java:84]: The same issue

File: src/main/java/practice/StreamPractice.java [hunk 7]
```diff
@@ -59,7 +80,12 @@ public List<Person> getWorkablePeople(int fromAge, int femaleToAge,
...
+                .map(Person::getCats)
+                .flatMap(Collection::stream)
+                .map(Cat::getName)
```
Comment (sarakhmen) [inline src/main/java/practice/StreamPractice.java:87]: think how you can reduce the number of map() calls

File: src/main/java/practice/StreamPractice.java [hunk 8]
```diff
@@ -75,6 +101,11 @@ public List<String> getCatsNames(List<Person> peopleList, int femaleAge) {
      * parametrized with Candidate in CandidateValidator.
      */
     public List<String> validateCandidates(List<Candidate> candidates) {
-        return Collections.emptyList();
+        CandidateValidator candidateValidator = new CandidateValidator();
+        return candidates.stream()
+                .filter(c -> candidateValidator.test(c))
```
Comment (sarakhmen) [inline src/main/java/practice/StreamPractice.java:106]:
```suggestion
        return candidates.stream()
                .filter(new CandidateValidator())
```
I'd recommend something like this, as this is more concise

**Review by sarakhmen — CHANGES_REQUESTED** [review]
Overall good job, but let's slightly improve your solution :)

File: src/main/java/practice/CandidateValidator.java (hunk 1 above)
Comment (karma-o) [reply src/main/java/practice/CandidateValidator.java:14]: remove it, we have a no param constructor available by default, if no other ones are defined

File: src/main/java/practice/StreamPractice.java (hunk 7 above)
Comment (karma-o) [reply src/main/java/practice/StreamPractice.java:87]: not fixed
```suggestion
                .flatMap(person -> person.getCats().stream())
                .map(Cat::getName)
```

File: src/main/java/practice/StreamPractice.java [hunk 9]
```diff
@@ -75,6 +101,10 @@ public List<String> getCatsNames(List<Person> peopleList, int femaleAge) {
      * parametrized with Candidate in CandidateValidator.
      */
     public List<String> validateCandidates(List<Candidate> candidates) {
-        return Collections.emptyList();
+        return candidates.stream()
+                .filter(c -> new CandidateValidator().test(c))
```
Comment (karma-o) [inline src/main/java/practice/StreamPractice.java:105]: not fixed, our Candidate validator is a Predicate itself so we can have it like this
```suggestion
                .filter(new CandidateValidator())
```

**Review by karma-o — CHANGES_REQUESTED** [review]

File: src/main/java/practice/CandidateValidator.java [hunk 10]
```diff
@@ -1,5 +1,28 @@
...
+
+    public CandidateValidator() {
+    }
+
+    @Override
+    public boolean test(Candidate candidate) {
+        String theBeginningOfPeriodInUkr = candidate.getPeriodsInUkr()
+                .split("-")[INDEX_FOR_FIRST_DATE];
+        String theEndOfPeriodInUrk = candidate.getPeriodsInUkr()
+                .split("-")[INDEX_FOR_SECOND_DATE];
```
Comment (karma-o) [inline src/main/java/practice/CandidateValidator.java:21]: not fixed, do not split the same string 2 times

**Review by karma-o — CHANGES_REQUESTED** [review]

**Review by karma-o — APPROVED** [review]

**Pull request closed without merging**